{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "%run age_predictors.py\n",
    "\n",
    "def load_data(metadata,datafile):\n",
    "\n",
    "    meta = pd.read_csv(metadata)\n",
    "    \n",
    "    dfo = pd.read_csv(datafile, header=None)\n",
    "    df = dfo.iloc[:,8:].T\n",
    "    pdat = df.iloc[:,0].asobject\n",
    "\n",
    "    uid = []\n",
    "    ageType = []\n",
    "    age = []\n",
    "    source = []\n",
    "    condition = []\n",
    "    ethnicity = []\n",
    "    gender = []\n",
    "    datep = []\n",
    "    whop = []\n",
    "    \n",
    "    #print dfo.shape\n",
    "    \n",
    "    for el in pdat:\n",
    "        tid = el.split('_')[0]\n",
    "        mdata = meta[meta['Tracking ID']==tid]\n",
    "        uid.append(el)  \n",
    "    \n",
    "        x = mdata['Age'].iloc[0].lower()\n",
    "        if 'fw' in x:\n",
    "            ageType.append('fw') \n",
    "            m = re.search('\\d{1,2} ?fw',x)\n",
    "            age.append(int(m.group(0)[:-2]))\n",
    "        elif 'nb' in x:\n",
    "            ageType.append('nb')\n",
    "            age.append(0)\n",
    "        elif 'da' in x:\n",
    "            ageType.append('nb')\n",
    "            age.append(0)\n",
    "        elif 'yr' in x:\n",
    "            ageType.append('yr')\n",
    "            m = re.search('\\d{1,3} ?yr',x)\n",
    "            # maybe use months sometime too, but not now\n",
    "            # m2 = re.search('\\d{1,2} ?mo',x) \n",
    "            age.append(int(m.group(0)[:-2]))\n",
    "        else:\n",
    "            print(\"OOOPS! \" + x)\n",
    "                \n",
    "\n",
    "        x = mdata['Condition'].iloc[0].lower()      \n",
    "#        could be any of these types:\n",
    "#                lab derived fibroblasts\n",
    "#                normal\n",
    "#                direct skin biopsy\n",
    "#                control for hgps\n",
    "#                relative of hgps\n",
    "#                control for ism\n",
    "#                ism\n",
    "#                ism(sorted)\n",
    "#                ism(hgps)\n",
    "#                ism(relative of hgps)\n",
    "#                heated direct skin biopsy\n",
    "        condition.append(x)\n",
    "        if 'fibroblast' in x:\n",
    "            source.append('fibroblast')\n",
    "        elif 'biopsy' in x:\n",
    "            source.append('biopsy')\n",
    "        else:\n",
    "            source.append('coriell')\n",
    "            \n",
    "        gender.append( mdata['Gender'].iloc[0].lower() )\n",
    "        ethnicity.append( mdata['Race/Ethnicity'].iloc[0].lower() )\n",
    "        oname = mdata['Original Name'].iloc[0]\n",
    "        datep.append( '20'+oname[:2]+'-'+oname[2:4]+'-'+oname[4:6] )\n",
    "        whop.append( oname.split('/')[1].split('_')[-1] )\n",
    "   \n",
    "            \n",
    "            \n",
    "    #\n",
    "    #df = df.drop(0)\n",
    "    #del df['patient data']\n",
    "    #print len(ageType),len(age), len(condition), len(source), len(df)\n",
    "    \n",
    "    \n",
    "    df.index = uid\n",
    "    df.drop(df.columns[0],axis=1,inplace=True)\n",
    "    df.columns = dfo.iloc[1:,0]\n",
    "    trnum=[]\n",
    "    for el in df.index:\n",
    "        xx = el.split('_')[0]\n",
    "        if xx=='pdl33': # get rid of the non-numeric tracking number which indicates #147 cells cultured to population doubling level 33\n",
    "            trnum.append(33147)\n",
    "        else:\n",
    "            trnum.append(int(xx))\n",
    "    df['tracking number'] = trnum\n",
    "    df['ethnicity'] = ethnicity\n",
    "    df['gender'] = gender\n",
    "    df['source'] = source\n",
    "    df['condition'] = condition\n",
    "    df['age type'] = ageType\n",
    "    df['age'] = age\n",
    "    df['date processed'] = datep\n",
    "    df['who processed'] = whop\n",
    "    \n",
    "    # rearrange so its easy to make a data matrix\n",
    "    cols = df.columns.tolist()\n",
    "    ncols = cols[-9:] + cols[:-9]\n",
    "    df = df[ncols]   \n",
    "\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "alldata = load_data('MetaDataSheet.csv','fpkm.csv')\n",
    "\n",
    "\n",
    "normals = (alldata.condition == 'normal') \n",
    "progeria = (alldata.condition == 'hgps')  \n",
    "genes = alldata.iloc[:,9:].astype(float) # get only the gene expression levels, exclude the metadata\n",
    "ages = alldata[normals].age \n",
    "age_prog = alldata[progeria].age\n",
    "\n"
   ],
   "outputs": [
    {
     "output_type": "error",
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pandas'",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "File \u001b[0;32m~/Desktop/Predicting-age-from-the-transcriptome-of-human-dermal-fibroblasts/age_predictors.py:9\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# let's be brutally honest here.  I wrote all this before I understood sklearn Pipelines.\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# if I had understood I could have saved myself a lot of trouble.\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# a really great student short project would be to replace the gene subsetting system with a custom\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# transformation and then just stick it in a Pipeline with whatever you want\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# JGF July 2021\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m__future__\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m division\n\u001b[0;32m----> 9\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mre\u001b[39;00m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01msklearn\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pandas'"
     ]
    },
    {
     "output_type": "error",
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [1]\u001b[0m, in \u001b[0;36m<cell line: 115>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    107\u001b[0m     df \u001b[38;5;241m=\u001b[39m df[ncols]   \n\u001b[1;32m    110\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m df\n\u001b[0;32m--> 115\u001b[0m alldata \u001b[38;5;241m=\u001b[39m \u001b[43mload_data\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mMetaDataSheet.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mfpkm.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    118\u001b[0m normals \u001b[38;5;241m=\u001b[39m (alldata\u001b[38;5;241m.\u001b[39mcondition \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnormal\u001b[39m\u001b[38;5;124m'\u001b[39m) \n\u001b[1;32m    119\u001b[0m progeria \u001b[38;5;241m=\u001b[39m (alldata\u001b[38;5;241m.\u001b[39mcondition \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhgps\u001b[39m\u001b[38;5;124m'\u001b[39m)  \n",
      "Input \u001b[0;32mIn [1]\u001b[0m, in \u001b[0;36mload_data\u001b[0;34m(metadata, datafile)\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mload_data\u001b[39m(metadata,datafile):\n\u001b[0;32m----> 5\u001b[0m     meta \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241m.\u001b[39mread_csv(metadata)\n\u001b[1;32m      7\u001b[0m     dfo \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(datafile, header\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m      8\u001b[0m     df \u001b[38;5;241m=\u001b[39m dfo\u001b[38;5;241m.\u001b[39miloc[:,\u001b[38;5;241m8\u001b[39m:]\u001b[38;5;241m.\u001b[39mT\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "alld2 = pd.read_csv('../raw data/fpkm-Sept1-removeISM.csv',header=None)\n",
    "#parse the gene symbol out of the full columns from raw HOMER output\n",
    "gene_symbol = alld2.iloc[1:,7].apply(lambda x: x.split('|')[0]) \n",
    "transcripts = alld2.iloc[1:,0]\n",
    "\n",
    "# make sure the raw data transcripts matches the transcripts from the usual fpkm table\n",
    "print(gene_symbol.shape, genes.columns.shape)\n",
    "print((transcripts == genes.columns).all())\n",
    "\n",
    "all_transcript_labelled_data = alldata.reset_index().set_index(['tracking number','age','condition','ethnicity','gender','source','date processed','who processed']).drop(['index','age type'],axis=1)\n",
    "all_gene_labelled_data = all_transcript_labelled_data.copy()\n",
    "all_gene_labelled_data.columns = gene_symbol # replace transcript labels with gene labels\n",
    "all_gene_labelled_data.to_csv('all_gene_labelled_data.csv')\n",
    "all_transcript_labelled_data.to_csv('all_transcript_labelled_data.csv')\n",
    "\n",
    "\n",
    "transcript_labelled_data = alldata[normals].reset_index().set_index(['tracking number','age','condition']).iloc[:,5:]\n",
    "gene_labelled_data = transcript_labelled_data.copy()\n",
    "gene_labelled_data.columns = gene_symbol\n",
    "gene_labelled_data.to_csv('gene_labelled_data.csv')\n",
    "transcript_labelled_data.to_csv('transcript_labelled_data.csv')\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.9 64-bit"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.9",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}